File: javascript/leetcode/515. Find Largest Value in Each Tree Row.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var largestValues = function (root) {
  const result = [];
  if (!root) return result;
  let seen = [];

  seen.push(root);
  let size = 1;
  let max = 0;

  while (1) {
    let added = 0;
    max = seen[0].val;
    while (size) {
      let removed = seen.shift();
      max = max >= removed.val ? max : removed.val;
      size--;

      if (removed.left) {
        added++;
        seen.push(removed.left);
      }

      if (removed.right) {
        added++;
        seen.push(removed.right);
      }
    }
    size = added;
    result.push(max);
    if (size === 0) break;
    max = seen[0].val;
  }

  return result;
};

File: c++/leetcode/515. Find Largest Value in Each Tree Row.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
  vector<int> largestValues(TreeNode *root)
  {
    vector<int> result;
    if (!root)
      return result;
    queue<TreeNode *> seen;
    seen.push(root);
    int size = 1;
    int max = 0;

    while (1)
    {
      int added = 0;
      max = seen.front()->val;
      while (size)
      {
        TreeNode *removed = seen.front();
        max = max >= removed->val ? max : removed->val;
        seen.pop();
        size--;

        if (removed->left)
        {
          added++;
          seen.push(removed->left);
        }

        if (removed->right)
        {
          added++;
          seen.push(removed->right);
        }
      }
      size = added;
      result.push_back(max);
      if (size == 0)
        break;
      max = seen.front()->val;
    }

    return result;
  }
};